dm/pkg/schema/tracker.go

```diff
@@ -414,7 +421,27 @@ func (tr *Tracker) CreateTableIfNotExists(table *filter.Table, ti *model.TableIn
 	return tr.dom.DDL().CreateTableWithInfo(tr.se, schemaName, ti, ddl.OnExistIgnore)
 }
 
-func (tr *Tracker) BatchCreateTableIfNotExist(tablesToCreate map[string]map[string]*model.TableInfo) error {
+func (tr *Tracker) RecreateTables(logCtx *tcontext.Context, tablesToDrop []*filter.Table, tablesToCreate map[string]map[string]*model.TableInfo) error {
+	tr.Lock()
+	defer tr.Unlock()
+	for _, tbl := range tablesToDrop {
+		// schema changed
+		if err := tr.DropTable(tbl); err != nil {
+			logCtx.L().Warn("failed to drop table from schema tracker",
+				zap.Stringer("table", tbl), log.ShortError(err))
+		}
+	}
+	for schemaName := range tablesToCreate {
+		// TODO: Figure out how to recover from errors.
+		if err := tr.CreateSchemaIfNotExists(schemaName); err != nil {
+			logCtx.L().Error("failed to rollback schema on schema tracker: cannot create schema",
+				zap.String("schema", schemaName), log.ShortError(err))
+		}
+	}
+	return tr.batchCreateTableIfNotExist(tablesToCreate)
```

Comment:
What will happen if the tracker failed to create the corresponding schema before creating the tables? Should we place this to the `err == nil` branch or let the tracker return error immediately after failing to create schemas?

Reply:
it's ignored before, copied from

https://github.com/pingcap/tiflow/blob/1a95ac2b5ce472048ecae418aa8b279c7f0d3e59/dm/syncer/checkpoint.go#L913-L920

will leave it this way now.

Reply:
Is it a bug or by design? @Ehco1996 @lance6716 🤔

Reply:
this error happens rarely since it's a embedded tidb, no connection issues, but if tracker's data directory is corrupted or disk full, may cause error. in that case it's hard to recover too, maybe that's why just add a todo.